src/Compiler/Compile.hs

```diff
@@ -435,43 +448,60 @@ compileProgram' term flags modules compileTarget fname program
-                                 isMainType tp = case expandSyn tp of
-                                                   TFun [] eff resTp  -> True -- resTp == typeUnit
-                                                   _                  -> False                                
-                             in case filter (isMainType . infoType) infos of
-                               [InfoFun{infoType=tp,infoRange=r}] 
-                                  -> do mbF <- checkUnhandledEffects flags loaded2 mainName r tp
-                                        case mbF of
-                                          Nothing -> return (Executable mainName tp, loaded2)
-                                          Just f  ->
-                                            let mainName2  = qualify (getName program) (newHiddenName "hmain")
-                                                expression = App (Var (if (isHiddenName mainName) then mainName -- .expr
-                                                                                                  else unqualify mainName -- main
-                                                                      ) False r) [] r
-                                                defMain    = Def (ValueBinder (unqualify mainName2) () (Lam [] (f expression) r) r r)  r Public (defFun []) InlineNever ""
-                                                program2   = programAddDefs program [] [defMain]
-                                            in do (loaded3,_) <- ignoreWarnings $ typeCheck loaded1 flags 0 coreImports program2
-                                                  return (Executable mainName2 tp, loaded3) -- TODO: refine the type of main2
-                               [info]
-                                  -> errorMsg (ErrorGeneral (infoRange info) (text "'main' must be declared as a function (fun)"))
-                               [] -> errorMsg (ErrorGeneral rangeNull (text "the type of 'main' must be a function without arguments" <->
-                                                                                      table [(text "expected type", ppType (prettyEnvFromFlags flags) mainType)
-                                                                                            ,(text "inferred type", ppType (prettyEnvFromFlags flags) (head (map infoType infos)))]))
-                               _  -> errorMsg (ErrorGeneral rangeNull (text "found multiple definitions for the 'main' function"))
-             Object -> return (Object,loaded2)
-             Library -> return (Library,loaded2)
-       
-       loaded4 <- liftIO $ codeGen term flags newTarget loaded3
+       (newTarget,loaded3) <- doCodeGen term flags loaded2 compileTarget program coreImports
+       (loaded4, outFile) <- liftIO $ case newTarget of
+            InMemory -> return (loaded3{loadedModule = (loadedModule loaded3){modOutputTime = Nothing}}, Nothing)
```

Comment:
I moved the evaluation part of `codeGen` up here, as it seems strange for `codeGen` to be in charge of evaluation. I'm not sure if it is best here (it could be even higher). Since we report the executable location, this is now possible to do here.